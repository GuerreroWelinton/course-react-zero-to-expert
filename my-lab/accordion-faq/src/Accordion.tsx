import AccordionItem from './accordion-item/AccordionItem';
import { questions } from './Accordion.constant';

import styles from './Accordion.module.css';

const Accordion = () => {
  return (
    <div data-testid="accordion" className={styles.container}>
      <h1>PREGUNTAS FRECUENTES</h1>
      {questions.map((item) => (
        <AccordionItem key={item.id} {...item} />
      ))}
    </div>
  );
};

export default Accordion;
