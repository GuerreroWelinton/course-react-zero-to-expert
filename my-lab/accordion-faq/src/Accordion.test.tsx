import { render, screen } from '@testing-library/react';
import { afterEach, describe, expect, test, vi } from 'vitest';
import { Accordion } from './Accordion';
import { QUESTIONS } from './Accordion.constant';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const mockComponent = vi.fn((_props: unknown) => <div data-testid="AccordionItem"></div>);

vi.mock('./accordion-item/AccordionItem', () => ({
  AccordionItem: (props: unknown) => mockComponent(props),
  // * Usarlo de la manera de abajo si exporto por defecto el componente
  // default: () => <div data-testid="AccordionItem"></div>,
}));

describe('Accordion', () => {
  afterEach(() => vi.clearAllMocks());

  test('should render question title and content', () => {
    const { container } = render(<Accordion />);

    const h1 = container.querySelector('h1');

    expect(h1?.innerHTML).toContain('PREGUNTAS');
  });

  test('should have correct CSS classes', () => {
    render(<Accordion />);

    const accordionClassName = screen.getByTestId('accordion').className;
    expect(accordionClassName).toContain('container');
  });

  test('should match snapshot', () => {
    const { container } = render(<Accordion />);

    expect(container).matchSnapshot();
  });

  test('should render the correct number of AccordionItem components', () => {
    render(<Accordion />);

    const accordionItems = screen.getAllByTestId('AccordionItem');

    expect(accordionItems.length).toBe(QUESTIONS.length);
  });

  test('should render AccordionItem with the correct props', () => {
    render(<Accordion />);

    expect(mockComponent).toHaveBeenCalledTimes(QUESTIONS.length);

    QUESTIONS.forEach((item) => expect(mockComponent).toHaveBeenCalledWith(item));
  });
});
