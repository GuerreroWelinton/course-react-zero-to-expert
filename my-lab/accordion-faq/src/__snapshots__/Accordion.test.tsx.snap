// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion > should match snapshot 1`] = `
<div>
  <div
    class="_container_b8a185"
    data-testid="accordion"
  >
    <h1>
      PREGUNTAS FRECUENTES
    </h1>
    <div
      class="_accordion_38baf4"
      data-testid="accordionItem"
    >
      <h1
        class="_question_38baf4"
      >
        1
        . 
        ¿Qué es React?
      </h1>
      <div
        class="_content_38baf4 _close_38baf4"
        data-testid="answer"
      >
        <p
          class="_answer_38baf4"
        >
          React es una biblioteca de JavaScript para construir interfaces de usuario. Te permite crear componentes reutilizables que gestionan su propio estado.
        </p>
      </div>
    </div>
    <div
      class="_accordion_38baf4"
      data-testid="accordionItem"
    >
      <h1
        class="_question_38baf4"
      >
        2
        . 
        ¿Qué es el JSX?
      </h1>
      <div
        class="_content_38baf4 _close_38baf4"
        data-testid="answer"
      >
        <p
          class="_answer_38baf4"
        >
          JSX es una extensión de la sintaxis de JavaScript que se parece a HTML. Permite escribir la estructura de tus componentes de una manera más declarativa y familiar.
        </p>
      </div>
    </div>
    <div
      class="_accordion_38baf4"
      data-testid="accordionItem"
      style="background-color: rgb(172, 199, 84);"
    >
      <h1
        class="_question_38baf4"
      >
        3
        . 
        ¿Qué hace el hook \`useState\`?
      </h1>
      <div
        class="_content_38baf4 _close_38baf4"
        data-testid="answer"
      >
        <p
          class="_answer_38baf4"
        >
          El hook \`useState\` permite a los componentes funcionales tener su propio estado. Devuelve un valor de estado y una función para actualizarlo, provocando un nuevo renderizado del componente.
        </p>
      </div>
    </div>
  </div>
</div>
`;
